<launch>
  <!-- Launch the Wanderer Node -->
  <node pkg="pathing" name="wanderer" type="wander" output="screen">
    <rosparam command="load" file="$(find pathing)/params.yaml"/>
  </node>

  <!-- Launch the Puppy Guarder Node -->
  <node pkg="pathing" name="puppy_guarder" type="puppy_guard" output="screen">
    <rosparam command="load" file="$(find pathing)/params.yaml"/>
  </node>

  <!-- Launch the Wall Follower Node -->
  <node pkg="pathing" name="patroller" type="wall_follow" output="screen">
    <rosparam command="load" file="$(find pathing)/params.yaml"/>
  </node>

  <!-- Launch the Gap Follower Node -->
  <node pkg="pathing" name="advancer" type="gap_follow" output="screen">
    <rosparam command="load" file="$(find pathing)/params.yaml"/>
  </node>
  
  <!-- Launch the Logger Node -->
  <node pkg="pathing" name="logger" type="log" output="screen">
    <rosparam command="load" file="$(find pathing)/params.yaml"/>
  </node>
  
  <group ns="navigation">
    <!-- Launch the Recaller Node -->
    <node pkg="pathing" name="retracer" type="recall" output="screen">
      <rosparam command="load" file="$(find pathing)/params.yaml"/>
    </node>

    <!-- Launch the Explorer Node -->
    <node pkg="pathing" name="explorer" type="explore" output="screen">
      <rosparam command="load" file="$(find pathing)/params.yaml"/>
    </node>
    
    <group ns="search">
      <!-- Launch the Searcher Node -->
      <node pkg="pathing" name="searcher" type="search" output="screen">
        <rosparam command="load" file="$(find pathing)/params.yaml"/>
      </node>
      
      <!-- Launch the Costmap generators -->
      <node pkg="pathing" name="remember_old_locations" type="time_costmap" output="screen">
        <rosparam command="load" file="$(find pathing)/params.yaml"/>
      </node>
      <node pkg="pathing" name="measure_route_convenience" type="ease_costmap" output="screen">
        <rosparam command="load" file="$(find pathing)/params.yaml"/>
      </node>
      <node pkg="pathing" name="analyze_unsearched_area" type="unsearched_area_costmap" output="screen">
        <rosparam command="load" file="$(find pathing)/params.yaml"/>
      </node>
      <node pkg="pathing" name="detect_bucket" type="object_costmap" output="screen">
        <rosparam command="load" file="$(find pathing)/params.yaml"/>
      </node>
      <node pkg="pathing" name="explore_the_unknown" type="frontier_costmap" output="screen">
        <rosparam command="load" file="$(find pathing)/params.yaml"/>
      </node>
    </group>
  </group>
</launch>
